README
Compiling the library if you have the svn sources (the easy way) ::

    ./autoclean.sh
    ./autogen.sh
    ./configure
    make

Optional (system installation) ::

    sudo make install



HISTORY
01.05.2011
    I got married the 16.04.2011 and know that now that difficult challenges are
    waiting me :

    1. Saving the file in place...
    2. Simplify the code and make it readable
    3. Do something smarter for the debugging outputs

02.01.2011
    Can't get anything that please me. Came into a dead end with buffers when
    trying to merge two delete buffers (we have to know which one is the first
    one in time because there is cases where the result might differ).

22.09.2010
    Now I'll try to do the documentation in fcl.h and other hearder files if
    any as these are distributed with the library.

21.09.2010
    The implemented interfaces are subject to change widely as I'm building
    things as the come to my mind.

07.09.2010
    This project has been started to cover a hole in the implementation of
    GtkHex. Someone on IRC (on the gnome french channel) told me that I could
    patch it in order to allow it to open files of any sizes but that in any
    cases I had to not break the API. The only way I found is to build a library
    on which GtkHex will rely on. This is supposed to be that library.
    The name of the project is a reference to the FCL Basket Club in Lyon.


test/libfcltest.c
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: t; c-basic-offset: 4 -*- */
/*
 *  libfcltest.c
 *  File Cache Library Test
 *
 *  (C) Copyright 2010 - 2012 Olivier Delhomme
 *  e-mail : olivier.delhomme@free.fr
 *  URL    : https://gna.org/projects/fcl/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or  (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY;  without even the implied warranty of
 *  MERCHANTABILITY  or  FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/** @file libfcltest.c
 */

#include "config.h"

#include <stdio.h>
#include <glib/gi18n-lib.h>

#include <fcl.h>

#include "libfcltest.h"

static void init_international_languages(void);

static void print_message(gboolean success, const char *format, ...);
static gchar *get_home_dir(void);

static guchar *fill_data_with_char(gint size, guchar car);

static void test_openning_and_closing_files(void);
static void test_openning_and_reading_files(void);
static void test_openning_and_overwriting_files(void);
static void test_openning_and_inserting_in_files(void);
static void test_openning_and_deleting_in_files(void);

/**
 *  Inits internationalisation
 */
static void init_international_languages(void)
{
    gchar *result = NULL;

    setlocale(LC_ALL, "");
    result = bindtextdomain(GETTEXT_PACKAGE, LOCALEDIR);

    fprintf(stdout, Q_("Gettext package : %s\n"), GETTEXT_PACKAGE);
    fprintf(stdout, Q_("Locale dir : %s\n"), LOCALEDIR);
    fprintf(stdout, Q_("Bindtextdomain : %s\n"), result);

    bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
    textdomain(GETTEXT_PACKAGE);
}


/**
 * Prints a message from a test
 * @param success indicates whether the test was sucessfull or not
 * @param format the message to print in a vprintf format
 * @param ... the va_list : list of arguments to fill in the format
 */
static void print_message(gboolean success, const char *format, ...)
{
    va_list args;
    gchar *str = NULL;

    va_start(args, format);
    str = g_strdup_vprintf(format, args);
    va_end(args);

    if (success)
        {
            fprintf(stdout, Q_("[ OK ] "));
            fprintf(stdout, "%s\n", str);
        }
    else
        {
            fprintf(stdout, Q_("[FAIL] "));
            fprintf(stdout, "%s\n", str);
        }
}


/**
 * Retrieves the home dir
 * @return the home dir in a gchar * that can be freed when no longer needed
 */
static gchar *get_home_dir(void)
{
    const char *homedir = NULL;

    homedir = g_getenv("HOME");

    if (homedir == NULL)
        {
            homedir = g_get_home_dir();
        }

    return g_strdup(homedir);
}


/**
 * Fills a buffer with a character
 * @param size : the size of the buffer that we want
 * @param car : character used to fill the buffer with
 * @return a guchar * buffer of the desired size filled with the desired
 *         character. It may be freed when no longer needed.
 */
static guchar *fill_data_with_char(gint size, guchar car)
{
    guchar *data = NULL;
    gint i = 0;

    data = (guchar *) g_malloc0(size * sizeof(guchar));

    for (i = 0; i < size; i++)
        {
            data[i] = car;
        }

    return data;
}


/**
 * This function tests openning files and closing them with different modes
 */
static void test_openning_and_closing_files(void)
{
    fcl_file_t *my_test_file = NULL;

    my_test_file = fcl_open_file("/bin/bash", LIBFCL_MODE_READ);
    if (my_test_file != NULL)
        {
            print_message(my_test_file != NULL, Q_("Opening a file in read mode (%Ld)"), my_test_file->real_size);
        }
    else
        {
            print_message(my_test_file != NULL, Q_("Opening a file in read mode."));
        }
    fcl_close_file(my_test_file, FALSE);

    my_test_file = fcl_open_file("/tmp/test.libfcl", LIBFCL_MODE_WRITE);
    print_message(my_test_file != NULL, Q_("Opening a file in write mode."));
    fcl_close_file(my_test_file, FALSE);

    my_test_file = fcl_open_file("/tmp/test.libfcl", LIBFCL_MODE_CREATE);
    print_message(my_test_file != NULL, Q_("Opening a file in create mode."));
    fcl_close_file(my_test_file, FALSE);

    my_test_file = fcl_open_file("/tmp/test_doesnotexists", LIBFCL_MODE_READ);
    print_message(my_test_file == NULL, Q_("Opening a file that does not exists in read only mode."));
    fcl_close_file(my_test_file, FALSE);

}


/**
 * This function test openning, reading in the file and closing them
 */
static void test_openning_and_reading_files(void)
{
    fcl_file_t *my_test_file = NULL;
    guchar *buffer = NULL;
    gsize size = 0;


    /* A valid test. Should return ELF as this is the magic number for a compiled /bin/bash */
    my_test_file = fcl_open_file("/bin/bash", LIBFCL_MODE_READ);
    size = 3;
    buffer = fcl_read_bytes(my_test_file, 1, &size);

    if (buffer != NULL)
        {
            print_message(buffer != NULL, Q_("Read (%d bytes at %d) : "), size, 1);
            fcl_print_data(buffer, 3, TRUE);
        }
    else
        {
            print_message(buffer != NULL, Q_("Reading 3 bytes in /bin/bash !"));
        }
    fcl_close_file(my_test_file, FALSE);


    /* This test will not return NULL anymore because I deleted the limit for
     * the global buffer (ie what may be claimed by an application
     */
    my_test_file = fcl_open_file("/bin/bash", LIBFCL_MODE_READ);
    size = LIBFCL_MAX_BUF_SIZE + 1;
    buffer = fcl_read_bytes(my_test_file, 1, &size);
    print_message(buffer != NULL, Q_("Openning a file in create mode and reading more than allowed."));
    fcl_close_file(my_test_file, FALSE);


    /* Reading data at the limits of the buffer */
    my_test_file = fcl_open_file("/bin/bash", LIBFCL_MODE_READ);
    size = 35;
    buffer = fcl_read_bytes(my_test_file, 65525, &size);
    if (buffer != NULL)
        {
            print_message(buffer != NULL, Q_("Read (%d bytes at %d) : "), size, 65525);
            fcl_print_data(buffer, size, TRUE);
        }
    else
        {
            print_message(buffer != NULL, Q_("Reading 35 bytes in /bin/bash !"));
        }
    fcl_close_file(my_test_file, FALSE);

    /* Reading data at the limits of the file */
    my_test_file = fcl_open_file("/bin/bash", LIBFCL_MODE_READ);
    size = 336;
    buffer = fcl_read_bytes(my_test_file, my_test_file->real_size - 336, &size);
    if (buffer != NULL)
        {
            print_message(buffer != NULL, Q_("Read (%d bytes at %d) : "), size, my_test_file->real_size - 336);
            fcl_print_data(buffer, size, TRUE);
        }
    else
        {
            print_message(buffer != NULL, Q_("Reading 336 bytes in /bin/bash !"));
        }
    fcl_close_file(my_test_file, FALSE);


    /* Reading data beyond the limits of the file */
    my_test_file = fcl_open_file("/bin/bash", LIBFCL_MODE_READ);
    size = 16384;
    buffer = fcl_read_bytes(my_test_file, my_test_file->real_size + 336, &size);
    if (buffer != NULL)
        {
            print_message(buffer == NULL, Q_("Read (%d bytes at %d) : "), size, my_test_file->real_size + 336);
            fcl_print_data(buffer, size, TRUE);
        }
    else
        {
            print_message(buffer == NULL, Q_("Reading %ld bytes in /bin/bash ! at %ld"), size, my_test_file->real_size + 336 );
        }
    fcl_close_file(my_test_file, FALSE);

}


/**
 * This function test openning, overwriting in files and closing them
 */
static void test_openning_and_overwriting_files(void)
{
    fcl_file_t *my_test_file = NULL;
    guchar *buffer = NULL;
    guchar *data = NULL;
    gboolean result = TRUE;
    gsize size = 0;
    gchar *filename = NULL;

    buffer = (guchar *) g_strdup_printf("ABC");

    /* This test tries to overwrite a readonly file */
    my_test_file = fcl_open_file("/bin/bash", LIBFCL_MODE_READ);
    size = 3;
    result = fcl_overwrite_bytes(my_test_file, buffer, 1, &size);

    print_message(result == FALSE, Q_("Trying to overwrite in a READ ONLY opened file"));

    fcl_close_file(my_test_file, FALSE);

    /* Testing to overwrite into a file (without saving) */
    filename = g_build_path(G_DIR_SEPARATOR_S, get_home_dir(), ".bashrc", NULL);
    my_test_file = fcl_open_file(filename, LIBFCL_MODE_WRITE);
    size = 3;
    result = fcl_overwrite_bytes(my_test_file, buffer, 2, &size);

    print_message(result == TRUE, Q_("Trying to overwrite in an opened file (%ld bytes)"), size);

    /* Verifiying that it worked out */
    size = 10;
    data = fcl_read_bytes(my_test_file, 0, &size);
    fcl_print_data(data, size, TRUE);

    fcl_close_file(my_test_file, FALSE);

    g_free(buffer);

}


/**
 * This function test openning, inserting bytes in files and closing them
 */
static void test_openning_and_inserting_in_files(void)
{
    fcl_file_t *my_test_file = NULL;
    guchar *buffer = NULL;
    guchar *data = NULL;
    gsize size = 0;

    buffer = (guchar *) g_strdup_printf("Is this inserted in the file ??");

    my_test_file = fcl_open_file("/tmp/createme", LIBFCL_MODE_CREATE);
    print_message(my_test_file != NULL, Q_("Opening a file in create mode."));

    /* Inserting bytes */
    fcl_insert_bytes(my_test_file, buffer, 0, 30);

    fprintf(stdout, "\nVerifying if everything is there !\n");
    /* Verifying this (double check here) */
    size = 200;
    data = fcl_read_bytes(my_test_file, 0, &size);
    fprintf(stdout, Q_("Size read : %d\n"), size);
    fcl_print_data(data, size, TRUE);

    fcl_close_file(my_test_file, FALSE);
}


/**
 * This function test openning, deleting bytes in files and closing them
 */
static void test_openning_and_deleting_in_files(void)
{
    fcl_file_t *my_test_file = NULL;
    guchar *data = NULL;
    gsize size = 0;
    gchar *filename = NULL;

    filename = g_build_path(G_DIR_SEPARATOR_S, get_home_dir(), ".bashrc", NULL);
    my_test_file = fcl_open_file(filename, LIBFCL_MODE_WRITE);
    print_message(my_test_file != NULL, Q_("Opening a file in create mode."));

    /* Deleting bytes */
    size = 30;
    fcl_delete_bytes(my_test_file, 5, &size);

    /* Verifying this */
    size = 200;
    data = fcl_read_bytes(my_test_file, 2, &size);
    fcl_print_data(data, size, TRUE);

    fcl_close_file(my_test_file, FALSE);
}




int main(int argc, char **argv)
{
    /* Initializing the locales */
    init_international_languages();

    /* Initializing the library */
    fprintf(stdout, Q_("\nNow testing libfcl ...\n"));
    libfcl_initialize();

    /**** Tests ****/
    fprintf(stdout, Q_("Testing opening and closing files :\n"));
    test_openning_and_closing_files();
    fprintf(stdout,"\n\n");

    fprintf(stdout, Q_("Testing opening and reading files :\n"));
    test_openning_and_reading_files();
    fprintf(stdout,"\n\n");

    fprintf(stdout, Q_("Testing opening and overwritting files :\n"));
    test_openning_and_overwriting_files();
    fprintf(stdout,"\n\n");

    fprintf(stdout, Q_("Testing opening and inserting in files :\n"));
    test_openning_and_inserting_in_files();
    fprintf(stdout,"\n\n");

    fprintf(stdout, Q_("Testing opening and deleting in files :\n"));
    test_openning_and_deleting_in_files();
    fprintf(stdout,"\n\n");


    return 0;
}



test/libfcltest.h
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: t; c-basic-offset: 4 -*- */
/*
 *  libfcltest.h
 *  File Cache Library Test Header
 *
 *  (C) Copyright 2010 - 2012 Olivier Delhomme
 *  e-mail : olivier.delhomme@free.fr
 *  URL    : https://gna.org/projects/fcl/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or  (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY;  without even the implied warranty of
 *  MERCHANTABILITY  or  FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/** @file libfcltest.h
 */

#ifndef _TESTLIBFCL_H_
#define _TESTLIBFCL_H_

#endif /* _TESTLIBFCL_H */


autogen.sh
#!/bin/sh

if [ -d m4 ]; then
    echo -n ""
else
    mkdir m4;
fi;

aclocal -I m4                          \
&& libtoolize                          \
&& automake --gnu --add-missing --copy \
&& autoconf
glib-gettextize --copy --force
intltoolize --copy --force --automake

if [ -e ./configure ]; then
    echo
    echo "You can now run ./configure"
    echo
else
    echo
    echo "Failure building the configure script."
    echo "You may miss required tools (aclocal 1.9, libtoolize, automake 1.9,"
    echo "autoconf 2.5x) to build fcl library."
    echo
    echo "Please install the appropriated package and re-run autogen.sh"
    echo "This might also happen if you don't have the m4 provided as a"
    echo "subdirectory in this package in the directory where aclocal will find"
    echo "them. Just copy ./m4/* in, usually, /usr/share/aclocal."
    echo
    echo "Or, it can also come from a configure.ac mistake (then it's my"
    echo "business ;)). Please report your issue on the fcl library mailing"
    echo "list."
    echo
fi


autoclean.sh
#!/bin/bash
#
# Author : Olivier Delhomme
# Licence : GNU GPL v2
#
if [ -f Makefile ];
 then
    make distclean
fi;
rm -f aclocal.m4
echo "rm -f aclocal.m4"
rm -fr autom4te.cache
echo "rm -fr autom4te.cache"
rm -f config.guess
echo "rm -f config.guess"
rm -f config.sub
echo "rm -f config.sub"
rm -f depcomp
echo "rm -f depcomp"
rm -f INSTALL
echo "rm -f INSTALL"
rm -f install-sh
echo "rm -f install-sh"
rm -f ltmain.sh
echo "rm -f ltmain.sh"
rm -f missing
echo "rm -f missing"
find . -name "Makefile" -exec rm -f {} \;
echo "find . -name Makefile -exec rm -f {} \;"
find . -name "Makefile.in" -exec rm -f {} \;
echo "find . -name Makefile.in -exec rm -f {} \;"
rm -f configure
echo "rm -f configure"
rm -f config.status config.log
echo "rm -f config.status config.log"
rm -f libtool
echo "rm -f libtool"
find . -name "*~" -exec rm -f {} \;
echo "find . -name *~ -exec rm -f {} \;"
find . -name "*#" -exec rm -f {} \;
echo "find . -name *# -exec rm -f {} \;"
find . -name *.gcd[ao] -exec rm -f {} \;
echo "find . -name *.gcd[ao] -exec rm -f {} \;"
find . -name *.gcov -exec rm -f {} \;
echo "find . -name *.gcov -exec rm -f {} \;"
find . -name *.gcno -exec rm -f {} \;
echo "find . -name *.gcno -exec rm -f {} \;"
find . -name gmon.out -exec rm -f {} \;
echo "find . -name gmon.out -exec rm -f {} \;"
rm -f src/libfcl.la
echo "rm -f src/libfcl.la"




src/include/fcl.h
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: t; c-basic-offset: 4 -*- */
/*
 *  fcl.h
 *  File Cache Library header file
 *
 *  (C) Copyright 2010 - 2012 Olivier Delhomme
 *  e-mail : olivier.delhomme@free.fr
 *  URL    : https://gna.org/projects/fcl/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or  (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY;  without even the implied warranty of
 *  MERCHANTABILITY  or  FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/** @file fcl.h
 * Header for the library
 *
 * This library is intented to manage the file at a level that allow one
 * to edit binary files directly without bothering with the memory issues
 * and such.
 *
 * @warning The library is certainly not thread safe !
 *
 */

#ifndef _LIBFCL_H_
#define _LIBFCL_H_

#include "config.h"

#include <glib.h>
#include <gio/gio.h>
#include <glib/gi18n-lib.h>
#include <stdio.h>
#include <string.h>

#include "config.h"

/**
 * @def LIBFCL_DATE
 * defines fcl's creation date
 *
 * @def LIBFCL_AUTHORS
 * defines fcl's library main authors
 *
 * @def LIBFCL_LICENSE
 * defines fcl library's license (at least GPL v2)
 *
 * @def LIBFCL_VERSION
 * defines fcl library's current version and release date of this version
 * (00.00.20XX means a development version)
 */
#define LIBFCL_AUTHORS "Olivier DELHOMME"
#define LIBFCL_DATE "07 09 2010"
#define LIBFCL_LICENSE N_("GPL v2 or later")
#define LIBFCL_VERSION "0.0.0 (00.00.0000)"


/**
 * @def LIBFCL_MODE_READ
 * Mode to open the file. In this mode, the file will be opened in read only
 * The library is not supposed to manage any buffers in this mode
 *
 * @def LIBFCL_MODE_WRITE
 * Mode to open the file. In this mode, the file is opened for writing and
 * reading (sort of appending). The file may be created if it does not exists.
 * The library will manage buffers and such.
 *
 * @def LIBFCL_MODE_CREATE
 * Mode to open a file. In this mode, the file is created. If an existing file
 * already exists it is replaced by the new one.
 */
#define LIBFCL_MODE_READ 0
#define LIBFCL_MODE_WRITE 2
#define LIBFCL_MODE_CREATE 4


/**
 * @struct fcl_file_t
 * Structure that contains all the definitions needed by the library for a
 * file.
 *
 * The sequence in fcl_file_t is ordered. The order is done with the offset of
 * the fcl_buf_t structure. The buffers in the sequence are modified buffers
 * only (at first at least)
 */
typedef struct
{
    gchar *name;                   /**< Name for the file                 */
    gint mode;                     /**< Mode in which the file was opened */
    goffset real_size;             /**< Actual size of the file           */
    GFile *the_file;               /**< The corresponding GFile           */
    GFileInputStream *in_stream;   /**< Stream used for reading           */
    GFileOutputStream *out_stream; /**< Stream used for writing           */
    GSequence *sequence;           /**< Sequence of buffers (fcl_buf_t)   */
} fcl_file_t;


/**
 * @struct fcl_buf_t
 * Structure that acts as a buffer
 */
typedef struct
{
    goffset offset;      /** Offset of the buffer (aligned with LIBFCL_BUF_SIZE) */
    goffset real_offset; /** Real offset of the buffer (calculated each time)    */
    gsize size;          /** Size of the buffer                                  */
    guchar *data;        /** The buffer (if any)                                 */
    gboolean in_seq;     /** Says wether the buffer is in the sequence or not    */
} fcl_buf_t;


/**
 * @struct fcl_stat_buf_t
 * Structure that can manage some statistics about the buffers in the sequence
 * of an fcl_file_t
 */
typedef struct
{
    gssize min_buf_size;   /** Minimum size of a buffer in the sequence */
    gssize max_buf_size;   /** Maximum size of a buffer in the sequence */
    gssize add_size;       /** Additions done within the sequence (file)*/
    gssize real_edit_size; /** Real size of the additions and deletions */
    guint64 n_bufs;        /** Number of buffers in the sequence        */
} fcl_stat_buf_t;


/**
 * @def LIBFCL_MAX_BUF_SIZE
 * Maximum buffer size that the library handles (This value is 2^20 as this was
 * the total amount of memory that my Atari 1040 ST had !). I don't know if this
 * is usefull in any manner... ?
 *
 * @def LIBFCL_BUF_SIZE
 * Default buffer size
 */
#define LIBFCL_MAX_BUF_SIZE 128 /* 1048576 */
#define LIBFCL_BUF_SIZE 8       /* 65536   */

/**
 * Public part of the library
 */

/**
 * This function initializes the library it has to be invoked first
 */
extern void libfcl_initialize(void);


/**
 * Opens a file. Nothing is performed on it.
 * @param path : the path of the file to be opened
 * @param mode : the mode to open the file (LIBFCL_MODE_READ, LIBFCL_MODE_WRITE,
 *               LIBFCL_MODE_CREATE).
 * @return a correctly filled fcl_file_t structure that represents the file
 */
extern fcl_file_t *fcl_open_file(gchar *path, gint mode);


/**
 * This function closes a fcl_file_t
 * @param the fcl_file_t to close
 * @param save : a gboolean to say wether if we want to save the file before
 *               closinf it or not.
 */
extern void fcl_close_file(fcl_file_t *a_file, gboolean save);


/**
 * This function reads an fcl_buf_t buffer from an fcl_file_t
 * @param a_file : the fcl_file_t file from which we want to read size bytes
 * @param position : the position where we want to read bytes
 * @param[in,out] size_pointer : the number of bytes we want to read. If this
 *                               value is higher than LIBFCL_MAX_BUF_SIZE the
 *                               function returns NULL. The value indicates the
 *                               real size of the data read hence, the real size
 *                               of the returned buffer
 * @return If everything is ok a filled guchar *buffer, NULL otherwise !
 */
extern guchar *fcl_read_bytes(fcl_file_t *a_file, goffset position, gsize *size_pointer);


/**
 * This function overwrites data in the file (size bytes of data at 'position'
 * in the file).
 * @warning it does do not writes to disk directly. It only overwrites correctly
 * the data into the file structure.
 * @param a_file : the fcl_file_t file to which we want to write size bytes
 * @param data : data to be overwritten in the file
 * @param position : position where to begin overwriting in the file
 * @param[in,out] size_pointer : size of the data to overwrite. returns the real
 *                               size of the data overwritten.
 * @return True If everything is ok, false otherwise
 */
extern gboolean fcl_overwrite_bytes(fcl_file_t *a_file, guchar *data, goffset position, gsize *size_pointer);


/**
 * This function inserts data in the file (size bytes of data at 'position'
 * in the file).
 * @warning it does do not writes to disk directly. It only inserts correctly
 * the data into the file structure.
 * @param a_file : the fcl_file_t file to which we want to write size bytes
 * @param data : data to be overwritten in the file
 * @param position : position where to begin overwriting in the file
 * @param size : size of the data to overwrite. As the insertion should never
 *               be less than size (or 0) one may consider that if the function
 *               retuned TRUE, size bytes were effectively inserted.
 * @return True If everything is ok, false otherwise
 */
extern gboolean fcl_insert_bytes(fcl_file_t *a_file, guchar *data, goffset position, gsize size);


/**
 * Deletes bytes in the buffers.
 * @param a_file : the fcl_file_t file to which we want to deleted size bytes
 * @param position : position where to delete size_pointer bytes in the file
 * @param[in,out] size_pointer : number of bytes to be deleted. Returns the
 * number of bytes effectively deleted (may be less than requested)
 * @return True if everything is ok, false otherwise.
 */
extern gboolean fcl_delete_bytes(fcl_file_t *a_file, goffset position, gsize *size_pointer);


/******************************************************************************/
/*********************************** Buffers **********************************/

/**
 * Prints a buffer data (exactly 'size' bytes)
 * @todo : print UTF8 encoded values
 * @param data : buffer data to be printed
 * @param size : number of bytes to prints (from data)
 * @param EOL : prints an End Of Line if TRUE, nothing if FALSE
 */
extern void fcl_print_data(guchar *data, gsize size, gboolean EOL);


/**
 * Gets the statistics of the buffers of a fcl_file_t file.
 * @param a_file : an openned fcl_file_t file.
 * @return A newly allocated fcl_stat_buf_t filled with the statistics about
 *         the sequence structure of the fcl_file_t structure. Returns NULL if
 *         the structure does not exists or does not have any buffers.
 */
extern fcl_stat_buf_t *fcl_get_buffer_stats(fcl_file_t *a_file);


/**
 * Inits the statistics structure to default values. The structure must be freed
 * when no longer needed
 * @return an newly allocated and default values populated fcl_stat_buf_t*
 *         structure
 */
extern fcl_stat_buf_t *fcl_init_buffer_stats();


/**
 * Prints stats of the buffers (if any) on an fcl_file_t file
 * @param an openned fcl_file_t*
 */
extern void fcl_print_buffer_stats(fcl_file_t *a_file);



#endif /* _LIBFCL_H_ */
